### DISCUSSION TITLE
Is it possible to add dynamic metadata during login into user_info/id_token

### DISCUSSION THREAD

**Discussion by johnlister** [zitadel/zitadel#8539] [Q&A]
Hi, I'm investigating using Zitadel for a project to replace our current IdP and we have a complicated login journey that offers a number of non standard (historical) auth methods and captures information during login that we relay to the clients via the user_info endpoint or id_token.

I was wondering if it is possible using the session API during the login journey to set add dynamic metadata (or any data) to the session, that is then available during the complement token flow/action to inject into the tokens, etc as custom claims.  The metadata relates to the authentication method used and may vary for each login based on user behaviour, so storing it in the user entity doesn't make sense.

Thanks

**Comment by hifabienne** [comment]
Hi @johnlister 

Our current hosted login is not implemented with the session api. And the complement token flow action is only available in the hosted login at the moment.
So at the moment it is not possible to mix those two up. 

In the future we will have a new hosted login on which we are already working (github.com/zitadel/typescript), and we will also add the action possiblity again there which is not yet implemented.

But if you are using the session api and you build your own login ui, you can implement all the custom flows you need to make your custom use cases possible. This is exactly why we implemented the session api, so you are able to implement all your custom needs.

does that already help you?

**Reply by johnlister** [reply]
Thanks for the quick reply. I'm looking at this page: https://zitadel.com/docs/guides/integrate/login-ui/oidc-standard and my current understanding of the session API, is as follows:
- use auth-request to create a "session"
- perform custom login steps and update session
- post session and auth-request to get redirect to RP

My challenge is how to get information from the above flow into the id_token for example. I'd assumed the actions applied during token generation/introspection regardless of the login method.

**Reply by hifabienne** [reply]
After checking internally with my colleagues, I need to change my answer. I had a missunderstanding in the flow.

You are right, you can use the complement token flow to add some custom claims also with the session API.

@stebenz @livio-a can you help me on this, is it possible to access the session and session metadata in that action?

**Comment by johnlister** [comment]
Thanks, out of curiosity can the SAML flows follow the same pattern using the session APIs and an external login UI ?
